File: frontend/lib/schemas/csa.ts
import { DocSchema } from "../doc-schema";

export const csaSchema: DocSchema = {
  id: "csa",
  title: "Cloud Service Agreement",
  subtitle: "Cover Page",
  infoBox:
    'This Cloud Service Agreement (the "CSA") consists of: (1) this Cover Page, (2) the Order Form, and (3) the Common Paper Cloud Service Agreement Standard Terms Version 2.0 identical to those posted at commonpaper.com/standards/cloud-service-agreement/2.0. Any modifications of the Standard Terms should be made on the Cover Page, which will control over conflicts with the Standard Terms.',
  coverFields: [
    {
      key: "productDescription",
      label: "Product Description",
      hint: "Description of the cloud service being provided",
      type: "textarea",
      placeholder: "Describe the cloud service...",
    },
    {
      key: "effectiveDate",
      label: "Effective Date",
      type: "date",
      placeholder: "Select a date",
    },
    {
      key: "subscriptionPeriod",
      label: "Subscription Period",
      hint: "Duration of the subscription",
      type: "text",
      placeholder: "e.g. 1 year from Effective Date",
    },
    {
      key: "fees",
      label: "Fees",
      hint: "Pricing and fee structure",
      type: "text",
      placeholder: "e.g. $1,000/month",
    },
    {
      key: "paymentProcess",
      label: "Payment Process",
      hint: "How fees will be collected",
      type: "text",
      placeholder: "e.g. Monthly invoicing, net 30",
    },
    {
      key: "governingLaw",
      label: "Governing Law",
      type: "text",
      placeholder: "State",
    },
    {
      key: "chosenCourts",
      label: "Chosen Courts",
      type: "text",
      placeholder: "City/county and state",
    },
  ],
  parties: [
    { label: "Provider", fields: ["name", "title", "company", "noticeAddress", "date"] },
    { label: "Customer", fields: ["name", "title", "company", "noticeAddress", "date"] },
  ],
  signingNote: "By signing this Cover Page, each party agrees to enter into this Cloud Service Agreement as of the Effective Date.",
  templateFile: "templates/CSA.md",
  footerNote: "Common Paper Cloud Service Agreement (Version 2.0) free to use under CC BY 4.0.",
  greeting:
    "Hi! I'm your Cloud Service Agreement assistant. I'll help you fill in the CSA Cover Page through conversation. This covers the key business terms for a cloud software or SaaS subscription. Let's start -- what cloud service or product is being provided? (e.g. 'Project management SaaS platform', 'Cloud-based CRM solution', 'Data analytics dashboard')",
  pdfFilename: "cloud-service-agreement.pdf",
};

File: backend/doc_configs/csa.py
from typing import Optional
from pydantic import BaseModel
from doc_configs.shared import PartialPartyInfo


SYSTEM_PROMPT = """You are a legal document assistant helping a user fill in a Cloud Service Agreement (CSA) Cover Page.

## Fields you must collect

1. **Product Description** -- description of the cloud service being provided
   Ask: "What cloud service or product is being provided? (e.g. 'Project management SaaS platform', 'Cloud-based CRM solution')"

2. **Effective Date** -- when the agreement takes effect
   Ask: "What is the effective date? (e.g. 'today', 'March 15 2026', 'next Monday')"
   For shortcuts like "today", "tomorrow", "Mon", "next Mon", use this lookup table:
{date_lookup}
   For full dates (e.g. "March 15 2026", "2026-03-15"), convert to YYYY-MM-DD format directly.

3. **Subscription Period** -- duration of the subscription
   Ask: "How long is the subscription period? (e.g. '1 year from Effective Date', '2 years', '36 months')"

4. **Fees** -- pricing and fee structure
   Ask: "What are the fees? (e.g. '$1,000/month', '$10,000/year', '$500/month per seat')"

5. **Payment Process** -- how fees will be collected
   Ask: "How will payments be made? (e.g. 'Monthly invoicing, net 30', 'Annual prepayment', 'Quarterly billing')"

6. **Governing Law** -- jurisdiction whose laws govern the agreement
   Ask: "Which state or jurisdiction's law should govern this agreement? (e.g. 'California', 'New York', 'Delaware')"

7. **Chosen Courts** -- which courts handle disputes
   Ask: "Which courts should have jurisdiction over disputes? (e.g. 'San Francisco, California', 'New York County, New York')"

8. **Party 1 (Provider)** -- name, title, company, notice address, signing date
   Ask for all sub-fields. Example: "Jane Smith, CEO of CloudCo, jane@cloudco.com, signing today"

9. **Party 2 (Customer)** -- same sub-fields as Party 1

## Current document state
{current_fields}

## Instructions
- Work through the fields in order, but adapt naturally to what the user already told you
- Ask 1-2 questions per turn maximum
- Always include a concrete example in parentheses when asking a question
- When the user answers, confirm what you captured then always end with a follow-up question
- Once all fields are collected, tell the user the document is complete and they can download the PDF

## Clearing fields
If the user asks to clear fields (e.g. "clear all", "reset", "clear the date"):
- To clear a specific field, return that field set to its empty value and confirm what was cleared
- To clear all fields, return every field reset to empty values, confirm, then ask the first question again
- Never clear silently

Return JSON with:
- "message": your conversational reply (include examples when asking questions)
- "fields": only the fields you are confident about from this turn; omit everything else
"""


class PartialFields(BaseModel):
    productDescription: Optional[str] = None
    effectiveDate: Optional[str] = None
    subscriptionPeriod: Optional[str] = None
    fees: Optional[str] = None
    paymentProcess: Optional[str] = None
    governingLaw: Optional[str] = None
    chosenCourts: Optional[str] = None
    party1: Optional[PartialPartyInfo] = None
    party2: Optional[PartialPartyInfo] = None


class AiResponse(BaseModel):
    message: str
    fields: PartialFields
